File: platform/leetcode/22/22.hs
-- https://leetcode.com/problems/generate-parentheses/description/

dfs :: Int -> Int -> [String]
dfs 0 0 = [[]]
dfs 0 rpr = map (')':) (dfs 0 (rpr-1))
dfs lpr 0 = map ('(':) (dfs (lpr-1) 0)
dfs lpr rpr = map ('(':) (dfs (lpr-1) rpr) ++
  if lpr < rpr then map (')':) (dfs lpr (rpr-1)) else []

generateParenthesis :: Int -> [String]
generateParenthesis n = dfs n n

main :: IO()
main = print (generateParenthesis 3)

File: platform/leetcode/22/22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def g(lpr, rpr, acc):
            match (lpr, rpr):
                case (0,0):
                    return [acc]
                case (0, _):
                    return g(0, rpr-1, acc+")")
                case _:
                    return g(lpr-1, rpr, acc+"(") + ([] if lpr >= rpr else g(lpr, rpr-1, acc+")"))
        return g(n,n,"")
